documentation/lets-make-a-plugin.md

```diff
@@ -0,0 +1,253 @@
+# Let's make a plugin!
+
+> Another example on how to use Got like a boss :electric_plug:
+
+Okay, so you already have learned some basics. That's great!
+
+When it comes to advanced usage, custom instances are really helpful.
+For example, take a look at [`gh-got`](https://github.com/sindresorhus/gh-got).
+It looks pretty complicated, but... it ain't.
+
+Before we start, we need to find the GitHub API docs: https://developer.github.com/v3/
+
+Let's write down the most important information:
+1. The root endpoint is `https://api.github.com/`.
+2. We will use the V3 API vesion.<br>
+   The `Accept` header needs to be set to `application/vnd.github.v3+json`.
+3. The body is in a JSON format.
+4. We will use OAuth2 for authorization.
+5. We may receive `400 Bad Request` or `422 Unprocessable Entity`.<br>
+   The body contains detailed information about the error.
+6. *Pagination?* Not yet. This is going to be a native feature of Got. We'll update this page accordingly when the feature is available.
+7. Rate limiting. These headers are interesting:
+
+- `X-RateLimit-Limit`
+- `X-RateLimit-Remaining`
+- `X-RateLimit-Reset`
+- `X-GitHub-Request-Id`
+
+Also `X-GitHub-Request-Id` may be useful.
+
+8. User-Agent is required.
+
+When we have all the necessary info, we can start mixing :cake:
+
+### The root endpoint
+
+Not much to do here, just extend an instance and provide `baseUrl` option:
+
+```js
+'use strict';
+const got = require('got');
+
+const instance = got.extend({
+	baseUrl: 'https://api.github.com'
+});
+
+module.exports = instance;
+```
+
+### v3 API
+
+GitHub needs to know which version we are using. We'll use the `Accept` header for that:
+
+```js
+const instance = got.extend({
+	baseUrl: 'https://api.github.com',
+	headers: {
+		accept: 'application/vnd.github.v3+json'
+	}
+});
+```
+
+### JSON body
+
+We'll use [`options.responseType`](../readme.md#responsetype):
+
+```js
+const instance = got.extend({
+	baseUrl: 'https://api.github.com',
+	headers: {
+		accept: 'application/vnd.github.v3+json'
+	},
+	responseType: 'json'
+});
+```
+
+### Authorization
+
+It's popular to set some environment variables, say `GITHUB_TOKEN`. You can modify the tokens in all your apps easily, right? Cool. What about... we want to provide a unique token for each app. Then we will need to create a new option - it will default to the environment variable, but you can easily override it.
```

Comment:
```suggestion
It's common to set some environment variables, for example, `GITHUB_TOKEN`. You can modify the tokens in all your apps easily, right? Cool. What about... we want to provide a unique token for each app. Then we will need to create a new option - it will default to the environment variable, but you can easily override it.
```